File: backend/Routes/CustomerProduct/CustomerProduct.js
const CustomerProduct = require('../../Models/CustomerProduct/CustomerProduct');
const express = require('express');
const router = express.Router();
const userAuth = require('../../Middleware/auth');

// Add Customer Product
router.post('/add', userAuth, async (req, res) => {
  try {
    const { customer, productCategory, manufacturer, slNumber, soldDate, warranty, amcStart, amcEnd } = req.body;

    if (warranty && (amcStart || amcEnd)) {
      return res.status(400).json({ message: 'Product is under warranty. Cannot add AMC now.' });
    }

    const productData = {
      customerId: customer,
      productCategoryId: productCategory,
      manufacturerId: manufacturer,
      slNumber,
      soldDate,
      warranty: warranty || null,
      amcStart: warranty ? null : (amcStart || null),
      amcEnd: warranty ? null : (amcEnd || null),
    };

    const newProduct = await CustomerProduct.create(productData);
    res.status(201).json({ message: 'Customer Product added successfully', product: newProduct });
  } catch (err) {
    res.status(500).json({ message: 'Error adding customer product', error: err.message });
  }
});

// Get by Customer
router.get('/by-customer/:customerId', userAuth, async (req, res) => {
    try {
      const products = await CustomerProduct.findAll({
        where: { customerId: req.params.customerId },
        order: [['createdAt', 'DESC']]
      });
      res.json(products);
    } catch (err) {
      res.status(500).json({ message: 'Error fetching customer products', error: err.message });
    }
  });

// Update
router.put('/:id', userAuth, async (req, res) => {
    try {
      const { customer, productCategory, manufacturer, slNumber, soldDate, warranty, amcStart, amcEnd } = req.body;
  
      if (warranty && (amcStart || amcEnd)) {
        return res.status(400).json({ message: 'Product is under warranty. Cannot add AMC.' });
      }
  
      const updateData = {
        customerId: customer,
        productCategoryId: productCategory,
        manufacturerId: manufacturer,
        slNumber,
        soldDate,
        warranty: warranty || null,
        amcStart: warranty ? null : (amcStart || null),
        amcEnd: warranty ? null : (amcEnd || null),
      };
  
      const [updated] = await CustomerProduct.update(updateData, { where: { id: req.params.id } });
      if (updated === 0) return res.status(404).json({ message: 'Customer Product not found' });
      const product = await CustomerProduct.findByPk(req.params.id);
      res.json({ message: 'Customer Product updated successfully', product });
    } catch (err) {
      res.status(500).json({ message: 'Error updating customer product', error: err.message });
    }
  });

// Delete
router.delete('/:id', userAuth, async (req, res) => {
    try {
      const deleted = await CustomerProduct.destroy({ where: { id: req.params.id } });
      if (deleted === 0) return res.status(404).json({ message: 'Customer Product not found' });
      res.json({ message: 'Customer Product deleted successfully' });
    } catch (err) {
      res.status(500).json({ message: 'Error deleting customer product', error: err.message });
    }
  });
  
module.exports = router;

File: frontend/lib/Models/product/customer_product.dart


class CustomerProduct {
  final String id;
  final Customer customer;
  final ProductCategory? productCategory;
  final Manufacturer? manufacturer;
  final String slNumber;
  final DateTime soldDate;
  final DateTime warranty;
  final DateTime? amcStart;
  final DateTime? amcEnd;
  final DateTime createdAt;
  final DateTime updatedAt;

  CustomerProduct({
    required this.id,
    required this.customer,
    this.productCategory,
    this.manufacturer,
    required this.slNumber,
    required this.soldDate,
    required this.warranty,
    this.amcStart,
    this.amcEnd,
    required this.createdAt,
    required this.updatedAt,
  });

  factory CustomerProduct.fromJson(Map<String, dynamic> json) {
    return CustomerProduct(
      id: (json['id'] ?? (json['id'] ?? (json['id'] ?? json['_id'])))?.toString() ?? '',
      customer: Customer.fromJson(json['customer']),
      productCategory: json['productCategory'] != null
          ? ProductCategory.fromJson(json['productCategory'])
          : null,
      manufacturer: json['manufacturer'] != null
          ? Manufacturer.fromJson(json['manufacturer'])
          : null,
      slNumber: json['slNumber'] ?? '',
      soldDate: DateTime.parse(json['soldDate']),
      warranty: DateTime.parse(json['warranty']),
      amcStart: json['amcStart'] != null ? DateTime.tryParse(json['amcStart']) : null,
      amcEnd: json['amcEnd'] != null ? DateTime.tryParse(json['amcEnd']) : null,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'customer': customer.toJson(),
      'productCategory': productCategory?.toJson(),
      'manufacturer': manufacturer?.toJson(),
      'slNumber': slNumber,
      'soldDate': soldDate.toIso8601String(),
      'warranty': warranty.toIso8601String(),
      'amcStart': amcStart?.toIso8601String(),
      'amcEnd': amcEnd?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class Customer {
  final String id;
  final String customerName;

  Customer({
    required this.id,
    required this.customerName,
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id: (json['id'] ?? (json['id'] ?? (json['id'] ?? json['_id'])))?.toString() ?? '',
      customerName: json['customerName'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'customerName': customerName,
    };
  }
}

class ProductCategory {
  final String id;
  final String productCategory;

  ProductCategory({
    required this.id,
    required this.productCategory,
  });

  factory ProductCategory.fromJson(Map<String, dynamic> json) {
    return ProductCategory(
      id: (json['id'] ?? (json['id'] ?? (json['id'] ?? json['_id'])))?.toString() ?? '',
      productCategory: json['productCategory'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'productCategory': productCategory,
    };
  }
}

class Manufacturer {
  final String id;
  final String manufacturer;

  Manufacturer({
    required this.id,
    required this.manufacturer,
  });

  factory Manufacturer.fromJson(Map<String, dynamic> json) {
    return Manufacturer(
      id: (json['id'] ?? (json['id'] ?? (json['id'] ?? json['_id'])))?.toString() ?? '',
      manufacturer: json['manufacturer'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'manufacturer': manufacturer,
    };
  }
}
class PostCustomerProductModel {
  final String customer;
  final String productCategory;
  final String slNumber;
  final String manufacturer;
  final DateTime soldDate;
  final DateTime warranty;
  final DateTime? amcStart;
  final DateTime? amcEnd;

  PostCustomerProductModel({
    required this.customer,
    required this.productCategory,
    required this.slNumber,
    required this.manufacturer,
    required this.soldDate,
    required this.warranty,
    this.amcStart,
    this.amcEnd,
  });

  Map<String, dynamic> toJson() {
    return {
      'customer': customer,
      'productCategory': productCategory,
      'slNumber': slNumber,
      'manufacturer': manufacturer,
      'soldDate': soldDate.toIso8601String(),
      'warranty': warranty.toIso8601String(),
      'amcStart': amcStart?.toIso8601String() ?? '',
      'amcEnd': amcEnd?.toIso8601String() ?? '',
    };
  }

  factory PostCustomerProductModel.fromJson(Map<String, dynamic> json) {
    return PostCustomerProductModel(
      customer: json['customer'] ?? '',
      productCategory: json['productCategory'] ?? '',
      slNumber: json['slNumber'] ?? '',
      manufacturer: json['manufacturer'] ?? '',
      soldDate: DateTime.parse(json['soldDate']),
      warranty: DateTime.parse(json['warranty']),
      amcStart: json['amcStart'] != null && json['amcStart'] != ''
          ? DateTime.tryParse(json['amcStart'])
          : null,
      amcEnd: json['amcEnd'] != null && json['amcEnd'] != ''
          ? DateTime.tryParse(json['amcEnd'])
          : null,
    );
  }
}
